nengo_spinnaker/utils/model.py

```diff
@@ -0,0 +1,380 @@
+    that lead to and from the operator.
+
+    Parameters
+    ----------
+    conn_map : `ConnectionMap`
+        Connection map from which the operator will be removed.  Note that the
+        connection map will be modified.
+    target : object
+        Operator to remove from the connection map.
+
+    Other Parameters
+    ----------------
+    force : bool
+        If False then the operator will only be optimised out if it is expected
+        that doing so will break network performance. If True (the default)
+        then the operator will be removed regardless.
+    """
+    # Grab the old connection map
+    old_conns = conn_map._connections.copy()
+    saved_conns = conn_map._connections.copy()
+
+    # Grab all the connections which are transmitted by the operator.
+    if target not in old_conns:
+        return True
+    out_conns = list(_get_port_kwargs(old_conns.pop(target)))
+
+    # Compute the most packets any object which receives packets from the
+    # operator will receive if the operator is not removed.
+    old_max_packets = _get_max_packets_received(out_conns)
+
+    # Prepare to compute the new equivalent of this value
+    new_rx = collections.defaultdict(lambda: 0)
+
+    # Create a new empty connection map dictionary and update the connection
+    # map to use the new dictionary
+    conns = collections.defaultdict(lambda: collections.defaultdict(list))
+    conn_map._connections = conns
+
+    # Copy across all connections from the old connection map; every time we
+    # encounter the operator as the sink of a signal we multiply that signal by
+    # each of the outgoing connections in turn and add those connections to the
+    # new connection map instead.
+    for kwargs in _iter_connection_map_as_kwargs(old_conns):
+        # The target will never appear as a source object as we removed it as a
+        # key before. Therefore determine what to do by looking at the sink
+        # object.
+        if kwargs["sink_object"] is target:
+            # If the sink object is the target then create multiply this
+            # connection with each of the outgoing signals for the target.
+            for new_kwargs in _multiply_signals(kwargs, out_conns):
+                conn_map.add_connection(**new_kwargs)
+
+                # Update the new number of packets that will be received by
+                # each sink (assuming that zeroed rows will be removed from the
+                # transform)
+                weight = np.sum(np.any(
+                    new_kwargs["transmission_parameters"].transform != 0.0,
+                    axis=1)
+                )
+                new_rx[new_kwargs["sink_object"]] += weight
+        else:
+            # Otherwise (re-)add this connection to the connection map.
+            conn_map.add_connection(**kwargs)
+
+    # Determine whether the changes made should be discarded.
+    new_max_packets = max(itervalues(new_rx))
+    discard_changes = new_max_packets > old_max_packets
```

Comment:
@tcstewar - this retains passthrough Nodes in cases where removing them would lead to an increase in the maximum number of packets received by any one of the objects connected to the passthrough Nodes. In principal this could be `discard_changes = (new_max_packets > old_max_packets) and (new_max_packets > some_safe_value)` so that it would retain passthrough Nodes only in cases where removing them worsened the number of packets above some threshold... I don't know what a sensible value for `some_safe_value` is yet.